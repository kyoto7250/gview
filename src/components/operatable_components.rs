use std::sync::{Arc, Mutex};

use crate::repository::RepositoryInfo;
use crossterm::event::KeyCode;
use ratatui::{layout::Rect, Frame};

// rust enum pass the operation command
pub enum Message {
    MultipleTimes(MultipleTimesOperation),
    Once(OnceOperation),
    NoAction,
    Error { message: String },
}

pub enum MultipleTimesOperation {
    Filtering {
        query: String,
    },
    SetUp {
        repository: Arc<Mutex<RepositoryInfo>>,
    },
    ChangeShowCommit,
}

pub enum OnceOperation {
    ShowFile { file: String },
    JumpToContentView,
    JumpToFiler,
}

#[derive(Clone, Copy, Debug)]
pub enum Focus {
    OFF,
    ON,
}

pub trait OperatableComponent {
    fn draw(&mut self, frame: &mut Frame, rect: Rect);
    fn process_focus(&mut self);
    fn process_events(&mut self, events: KeyCode) -> Message;
    fn handle_message(&mut self, message: &Message) -> Message;
}
