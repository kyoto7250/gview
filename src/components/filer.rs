use std::cmp::min;

use crossterm::event::KeyCode;
use ratatui::{
    layout::Rect,
    style::{Color, Style},
    widgets::{Block, Borders, List, ListItem, ListState},
    Frame,
};

use super::operatable_components::{
    Focus, Message, MultipleTimesOperation, OnceOperation, OperatableComponent,
};

pub struct Filer {
    focus: Focus,
    selected: usize,
    query: String,
    items: Vec<String>,
    results: Vec<String>,
}

impl Filer {
    pub fn new() -> Self {
        Self {
            focus: Focus::OFF,
            selected: 0,
            query: "".to_owned(),
            items: vec![],
            results: vec![],
        }
    }

    fn _handle_message(&mut self, message: &Message) -> Message {
        match message {
            Message::Once(OnceOperation::JumpToContentView) => self.focus = Focus::OFF,
            Message::Once(OnceOperation::SetUp { repository }) => {
                let binding = repository.clone();
                let mut repository = binding.lock().unwrap();
                let items = repository.recursive_walk().unwrap();
                self.items = items.clone();
                self.results = items;
            }
            Message::MultipleTimes(MultipleTimesOperation::Filtering { query }) => {
                self.focus = Focus::ON;
                self.query = query.to_owned();
                self.results = self
                    .items
                    .clone()
                    .into_iter()
                    .filter(|item| self.query.is_empty() || item.contains(&self.query))
                    .collect();

                if self.results.is_empty() {
                    self.results.push("not found".to_owned())
                }

                self.selected = min(self.selected, self.results.len().saturating_sub(1));
                return Message::Once(OnceOperation::ShowFile {
                    file: self.results[self.selected].to_owned(),
                });
            }
            _ => {}
        }
        Message::NoAction
    }
}

impl<'a> OperatableComponent for Filer {
    fn draw(&mut self, frame: &mut Frame, rect: Rect) {
        let list_items: Vec<ListItem> = self
            .results
            .iter()
            .map(|item| ListItem::new(item.to_owned()))
            .collect();

        let list = List::new(list_items)
            .block(Block::default().borders(Borders::NONE))
            .highlight_symbol(">> ")
            .style(match self.focus {
                Focus::ON => Style::default(),
                Focus::OFF => Style::default().fg(Color::DarkGray),
            });

        let mut list_state = ListState::default();
        list_state.select(Some(self.selected));
        frame.render_stateful_widget(list, rect, &mut list_state);
    }

    fn process_focus(&mut self) {
        match self.focus {
            Focus::OFF => self.focus = Focus::ON,
            Focus::ON => self.focus = Focus::OFF,
        }
    }
    fn process_events(&mut self, code: KeyCode) -> Message {
        match code {
            KeyCode::Up => {
                if self.selected > 0 {
                    self.selected -= 1;
                    return Message::Once(OnceOperation::ShowFile {
                        file: self.results[self.selected].clone(),
                    });
                }
            }
            KeyCode::Down => {
                if self.selected + 1 <= self.results.len().saturating_sub(1) {
                    self.selected += 1;
                    return Message::Once(OnceOperation::ShowFile {
                        file: self.results[self.selected].clone(),
                    });
                }
            }
            KeyCode::Enter => return Message::Once(OnceOperation::JumpToContentView),
            _ => {}
        }
        Message::NoAction
    }

    fn handle_message(&mut self, message: &Message) -> Message {
        // allow pattern
        // 1. MultipleTimes -> Once
        // 2. MultipleTimes -> NoAction
        // 3. Once -> NoAction
        // 4. NoAction -> NoAction
        return match (message, self._handle_message(message)) {
            (Message::MultipleTimes(_), Message::MultipleTimes(_)) => unreachable!(),
            (Message::Once(_), Message::MultipleTimes(_)) => unreachable!(),
            (Message::Once(_), Message::Once(_)) => unreachable!(),
            (Message::NoAction, Message::MultipleTimes(_)) => unreachable!(),
            (Message::NoAction, Message::Once(_)) => unreachable!(),
            (_, new_message) => new_message,
        };
    }
}
