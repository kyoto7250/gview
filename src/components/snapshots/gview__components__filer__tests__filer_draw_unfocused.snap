---
source: src/components/filer.rs
expression: "format!(\"{:?}\", buffer)"
snapshot_kind: text
---
Buffer {
    area: Rect { x: 0, y: 0, width: 80, height: 24 },
    content: [
        "┌3 files───────────────────────────────────────────────────────────────────────┐",
        "│>> src/main.rs                                                                │",
        "│   src/lib.rs                                                                 │",
        "│   README.md                                                                  │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "└──────────────────────────────────────────────────────────────────────────────┘",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 1, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 1, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 2, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 2, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 3, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 3, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 4, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 4, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 5, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 5, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 6, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 6, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 7, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 7, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 8, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 8, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 9, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 9, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 10, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 10, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 11, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 12, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 12, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 13, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 13, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 14, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 15, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 15, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 16, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 17, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 17, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 18, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 19, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 19, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 20, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 20, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 21, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 21, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 22, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 79, y: 22, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
